Created mockup Schema fixes #17

nimeshnayaju opened the pull request:
…ade a query that returned the document

nimeshnayaju commented:
Schema creation on models.js and controller code on controller-test.js (look at commit -> file changed)

evscott commented:
Will review this later.

evscott commented on controller-test.js:
```diff
@@ -0,0 +1,33 @@
+//const mongoose = require('mongoose');
```
Good ideas here. This feature should be tested, however, could you refactor these tests to be unit tests using [Jest](https://jestjs.io/)?

evscott commented on models.js:
```diff
@@ -0,0 +1,19 @@
+const mongoose = require('mongoose');
```
Looks good.

evscott commented:
Also, for future pull requests, to try keep your initial 'commit' message short. As you can see, it's used as the title which has length restrictions, and any text that exceeds this restriction spills over into the summary. The best practice is to use a descriptive title as an initial commit message, and then summarize the changes in the PR's summary.

nimeshnayaju commented:
Noted! I will look more into Jest and creating a unit test. Thanks for the feedback!

nimeshnayaju commented:
I used Mocha for the time being because of my familiarity with the framework. Refactored the controller code and made separate directories for tests and models. Created test suite for GET request for individual products using id. Used express for handling the routes.

nimeshnayaju commented:
In order to run the test, type the following command into your terminal:

node run test

nimeshnayaju commented:
It's not giving an error when I run it on my machine, but the Travics CI is throwing a UnhandledPromiseRejectionWarning Exception

ShekharDewan commented:
I suspect that was happening because there wasn't any actual server to query and it tried to? Let's hope the changes you made fix it. And the ID part might change as we get more specifics on the barcode and what wikidata sends back, but I think the design looks fine for now. Good stuff!

evscott commented on package.json:
```diff
@@ -5,7 +5,8 @@
   "scripts": {
     "start": "nodemon server.js",
-    "test": "jest",
+    "test": "mocha tests/**/*.test.js",
```
Because we haven't begun using Jest yet, I'm personally fine with moving to Mocha. The rest of the team should have a say on this decision though.

ShekharDewan replied:
And in case the move to mocha was made to attempt to get the build to succeed, I think we can just revert the change.

ShekharDewan replied:
Yeah I think unless there is a good reason to use mocha, we should stick to jest.

evscott commented on tests/controller.test.js:
```diff
@@ -0,0 +1,66 @@
+});
+
+// Test suite
+describe('GET /products/:id', () => {
```
Nice test cases.

evscott commented on controller.js:
```diff
@@ -11,4 +11,55 @@ let ping = async (req, res) => {
 
```
In this file we shouldn't be creating a new Express instance. That should only be done in the root file, which passes its Express object into the routes module which appends controllers to specific routes to handle requests. Those handles should be unit tested.

nimeshnayaju replied:
Noted. Will make amendments in the next push. Thanks for the feedback

evscott commented:
Good ideas, but some changes are needing to be made. 

1. Travis should use a stable version of node, not the latest.
2. The 'beforeEach' clause in our Mocha tests needs to be fixed - see Travis's build error.
3. The controller should only consists of functions which handle requests. See the relationship between our express object `app` in our `server.js` file, our `routes` module, and `ping` in `controller.js`.

nimeshnayaju commented:
I made some changes from within the repo, which was not relevant to my newer push. Opted for force-push for the reason. Will try to curb on that in the next commits.

nimeshnayaju commented:
Just realized, I misspelled 'services' in my code. I can see the test failing again for that reason.

nimeshnayaju commented:
The test has successfully passed, but the timeout exception is being shown because the mongoose keeps the connection open even after all the tests has passed. Altering the state of external entity via unit tests is not recommended. I will let Elliot have his say on this.

nimeshnayaju commented:
> Good ideas, but some changes are needing to be made.
> 
> 1. Travis should use a stable version of node, not the latest.
> 2. The 'beforeEach' clause in our Mocha tests needs to be fixed - see Travis's build error.
> 3. The controller should only consists of functions which handle requests. See the relationship between our express object `app` in our `server.js` file, our `routes` module, and `ping` in `controller.js`.

PASSED!!!!!

nimeshnayaju commented:
Refactored controller.js, and added route folder to handle requests for individual route. Added some documents via Robo 3T and made GET request from Postman; the request and route handling worked perfectly.
